cookbooks/fb_storage/resources/format_devices.rb

```diff
@@ -0,0 +1,159 @@
+    )
+    msg = []
+    msg << "Partitioning #{to_do[:devices]}" unless to_do[:devices].empty?
+    msg << "Formatting #{to_do[:partitions]}" unless to_do[:partitions].empty?
+    converge_by msg.join(', ') do
+      converge_storage(to_do, storage)
+    end
+  end
+
+  # No matter what, we pass the data onto fb_fstab - this isn't a change
+  # fb_fstab will report any changes it makes when it runs.
+  #
+  # unless we're in firstboot_os, since we don't converge storage until we're in
+  # firstboot_tier
+  unless node.firstboot_os?
+    storage.gen_fb_fstab(node).each do |name, data|
+      node.default['fb_fstab']['mounts'][name] = data
+    end
+  end
+
+  # Determine if the kernel has multi-queue support enabled
+  kernel_ver = FB::Version.new(node['kernel']['release'])
+  kernel_mq_ver = FB::Version.new('4.11')
+  kernel_has_mq = kernel_ver >= kernel_mq_ver
+
+  # Finally, set any tunables
+  storage.config.each_key do |device|
+    dev = FB::Storage.device_name_from_path(device)
+
+    if node['fb_storage']['tuning']['scheduler'] # ~FC023
+      fb_sysfs "/sys/block/#{dev}/queue/scheduler" do
+        # Kernels prior to 4.11 do not have multi-queue support - t19377518
+        not_if { dev.start_with?('nvme') && !kernel_has_mq }
+        type :list
+        value node['fb_storage']['tuning']['scheduler']
+      end
+    end
+
+    if node['fb_storage']['tuning']['queue_depth'] # ~FC023
+      fb_sysfs "/sys/block/#{dev}/device/queue_depth" do
+        type :int
+        value node['fb_storage']['tuning']['queue_depth']
+      end
+    end
+
+    if node['fb_storage']['tuning']['discard_max_bytes'] # ~FC023
+      fname = "/sys/block/#{dev}/device/discard_max_bytes"
+      fb_sysfs fname do
+        only_if do
+          # Only enables this setting when file exists and
+          # file has root write permission bit set
+          # This is due to older kernels don't support any
+          # change to this file, we can tell if this is supported
+          # by checking if root write permission bit is set
+          # We're using the bitwise operation because internally ruby uses
+          # eaccess which seems to consider root able to write anything
+          # regardless of mode bits.
+          ::File.exist?(fname) && ::File.stat(fname).mode & 0200 == 128
+        end
+        type :int
+        value node['fb_storage']['tuning']['discard_max_bytes']
+      end
+    end
+
+    # put a hard maximum on max_sectors_kb for nvme devices by default
+    # (see T22006954)
+    ignore_failure = FB::Fstab.get_in_maint_disks.include?(dev)
+    nvme_max_sectors = 8192
+    max_sectors_kb =
+      node['fb_storage']['tuning']['max_sectors_kb']
+    begin
+      max_hw_sectors_kb =
+        ::File.read("/sys/block/#{dev}/queue/max_hw_sectors_kb").to_i
+    rescue StandardError
+      if ignore_failure
+        next
+      else
+        raise
+      end
+    end
+
+    if dev.start_with?('nvme')
```

Comment:
... speaking of nvme...

Reply:
Yeah, there's a bunch of enablement logic here and there that's currently only used by internal-only handlers. I figured it was still generic enough to leave around.